### PR TITLE
Add support to log customized tags to runs created by autolog

### LINKED ISSUE #9064: [FR] Support logging customized tags to runs created by autolog
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

We'd like to add some customized tags to the runs created by autologging. For example, it would be great if autolog could support an additional parameter `tags` like `mlflow.autolog(tags={"custom_tag": "auto_created_run"})` and log the given tags to all runs created by autolog.

### Motivation

> #### What is the use case for this feature?
In hyperparameter tuning and AutoML scenarios, a lot of runs will be created. Currently, we couldn't add  customized tags to all the runs created by autolog easily.

> #### Why is this use case valuable to support for MLflow users in general?
Hyperparameter tuning and AutoML are broadly used by data science and machine learning communities, many of the engineers are also MLFlow users.

> #### Why is this use case valuable to support for your project(s) or organization?
By adding customized tags, we could organize the runs more flexibly.

> #### Why is it currently difficult to achieve this use case?
Currently, there is no such a parameter to pass into the autolog functions. We can start a run with tags and run training code in the context of the run, but we need to start hunderds of runs manually for each of the training code.

### Details

We'd like to add some customized tags to the runs created by autologging. For example, it would be great if autolog could support an additional parameter `tags` like below:
```
import mlflow
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_diabetes
from sklearn.ensemble import RandomForestRegressor


mlflow.autolog(tags={"custom_tag": "auto_created_run"})

db = load_diabetes()
X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

# Create and train models.
rf = RandomForestRegressor(n_estimators=100, max_depth=6, max_features=3)
rf.fit(X_train, y_train)

# Create and train models.
rf1 = RandomForestRegressor(n_estimators=20, max_depth=6, max_features=3)
rf1.fit(X_train, y_train)
```
We want the tags be like:
image

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by thinkall**
## Related Issues/PRs

Resolve #9064

## What changes are proposed in this pull request?

Added a parameter `extra_tags` to all `autolog` functions, and passed it to `safe_patch` function in `mlflow.utils.autologging_utils.safety.py`.

With this `extra_tags`, users can log customized tags to runs created by autolog.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users can set the parameter `extra_tags` in `mlflow.autolog(extra_tags={"test_tag": "autolog"})` and all the other specific autolog  functions such as `mlflow.sklearn.autolog(extra_tags={"test_tag": "sklearn_autolog"})` to log the `extra_tags` into runs created by autolog.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by thinkall** [comment]
Hi @serena-ruan , could you help review the PR? Thanks.

**Review by serena-ruan — COMMENTED** [review]
@thinkall Overall looks great! Left a few comments :)

I think it could be more clear that we split tags test.

**Comment by thinkall** [comment]
> @thinkall Overall looks great! Left a few comments :)
> 
> I think it could be more clear that we split tags test.

Thank you Serana, I've split all the tests for extra_tags. Warning messages and built-in tags checking have been included as well.

File: mlflow/utils/autologging_utils/safety.py [hunk 1]
```diff
@@ -304,14 +309,23 @@ class definition or a function object. If it is a function object, the
...
 
     if manage_run:
+        tags = {MLFLOW_AUTOLOGGING: autologging_integration}
+        if isinstance(extra_tags, dict):
+            if MLFLOW_AUTOLOGGING in extra_tags:
+                extra_tags.pop(MLFLOW_AUTOLOGGING)
+                _logger.warning("Tag %s is being ignored because it conflicts with a built-in tag.", MLFLOW_AUTOLOGGING)
+            tags.update(extra_tags)
+        elif extra_tags is not None:
+            _logger.warning("The parameter `extra_tags` is being ignored because it is expected to be a dictionary, but the provided value is %s", type(extra_tags))
```
Comment (serena-ruan) [inline mlflow/utils/autologging_utils/safety.py]:
```suggestion
        tags = {MLFLOW_AUTOLOGGING: autologging_integration}
        if extra_tags:
            if isinstance(extra_tags, dict):
                if MLFLOW_AUTOLOGGING in extra_tags:
                    extra_tags.pop(MLFLOW_AUTOLOGGING)
                    _logger.warning(f"Tag `{MLFLOW_AUTOLOGGING}` is ignored as it is a reserved tag by MLflow autologging.")
                tags.update(extra_tags)
            else:
                raise MlflowException.invalid_parameter_value(f"Invalid `extra_tags` type: expecting dictionary, received `{type(extra_tags).__name__}`")
```

Sorry I think it makes more sense to throw an exception if it's not a dictionary :)

File: tests/fastai/test_fastai_autolog.py [hunk 2]
```diff
@@ -84,6 +84,15 @@ def test_fastai_autolog_ends_auto_created_run(iris_data, fit_variant):
     assert mlflow.active_run() is None
 
 
+def test_extra_tags_fastai_autolog(iris_data):
+    mlflow.fastai.autolog(extra_tags={"test_tag": "fastai_autolog"})
+    model = fastai_tabular_model(iris_data)
+    model.fit_one_cycle(1)
+    
+    run = mlflow.last_active_run()
+    assert run.data.tags["test_tag"] == "fastai_autolog"
```
Comment (serena-ruan) [inline tests/fastai/test_fastai_autolog.py:93]:
```suggestion
    assert run.data.tags["test_tag"] == "fastai_autolog"
    assert run.data.tags[MLFLOW_AUTOLOGGING] == "fastai"
```

Comment (serena-ruan) [reply tests/fastai/test_fastai_autolog.py:93]: Can we add MLFLOW_AUTOLOGGING tag check for all flavors as well? Since you've already add tags check :)

Comment (thinkall) [reply tests/fastai/test_fastai_autolog.py:93]: Added :)

**Comment by serena-ruan** [comment]
LGTM! @harupy You wanna take a look? 😃

**Comment by thinkall** [comment]
> 

Thank you @serena-ruan, and hi, @harupy ! Just a gentle reminder to take a look :-)

**Review by serena-ruan — APPROVED** [review]
LGTM

**Comment by tiagodavi** [comment]
This is not working with langchain autolog.. mlflow: 3.5.1

```
 mlflow.autolog(extra_tags={"tracking.version": "1.0.0", "tracking.evaluated": "NO", "tracking.date": today})
 mlflow.langchain.autolog()
 mlflow.set_experiment("Tracking")
```

It creates traces without the extra_tags.

**Pull request merged**